Ruby/Ruby.sublime-syntax

```diff
@@ -203,8 +203,70 @@ contexts:
       scope: constant.language.ruby
     - match: '\b(__(FILE|LINE|ENCODING)__|self)\b(?![?!])'
       scope: variable.language.ruby
-    - match: '\b(0[xX]\h(_?\h)*|\d(_?\d)*(\.(?![^[:space:][:digit:]])(_?\d)*)?([eE][-+]?\d(_?\d)*)?|0[bB][01]+)\b'
-      scope: constant.numeric.ruby
+    # numbers with 'ir' suffix are invalid
+    - match: '\b(?:\d(?:_?\d)*(?:\.\d(?:_?\d)*)?(?:[eE][-+]?\d(?:_?\d)*)?|0[dDxXoObB]?\h(?:_?\h)*)ir\b'
+      scope: invalid.illegal.ruby
+    - match: '\b(0[dD])\d(?:_?\d)*(r)?(i)?\b'
+      scope: constant.numeric.integer.decimal.ruby
+      captures:
+        1: punctuation.definition.numeric.decimal.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    - match: '\b(0[xX])\h(?:_?\h)*(r)?(i)?\b'
+      scope: constant.numeric.integer.hexadecimal.ruby
+      captures:
+        1: punctuation.definition.numeric.hexadecimal.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    - match: '\b(0[oO]?)[0-7](?:_?[0-7])*(r)?(i)?\b'
+      scope: constant.numeric.integer.octal.ruby
+      captures:
+        1: punctuation.definition.numeric.octal.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    - match: '\b(0[bB])[01](?:_?[01])*(r)?(i)?\b'
+      scope: constant.numeric.integer.binary.ruby
+      captures:
+        1: punctuation.definition.numeric.binary.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    # 1.1e1i, 1e1i
+    - match: '\b\d(?:_?\d)*(?:(\.)\d(?:_?\d)*)?[eE][-+]?\d(?:_?\d)*(i)\b'
+      scope: constant.numeric.complex.imaginary.ruby
+      captures:
+        1: punctuation.separator.decimal.ruby
+        2: storage.type.numeric.imaginary.ruby
+    # 1.1ri, 1.1i, 1ri, 1i
+    - match: '\b\d(?:_?\d)*(?:(\.)\d(?:_?\d)*)?(r)?(i)\b'
+      scope: constant.numeric.complex.imaginary.ruby
+      captures:
+        1: punctuation.separator.decimal.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    # 1.i
+    - match: '\b\d(?:_?\d)*(\.)(i)\b'
+      scope: constant.numeric.complex.imaginary.ruby
+      captures:
+        1: punctuation.separator.decimal.ruby
+        2: storage.type.numeric.imaginary.ruby
+    # 1.1r, 1r
+    - match: '\b\d(?:_?\d)*(?:(\.)\d(?:_?\d)*)?(r)\b'
+      scope: constant.numeric.rational.ruby
+      captures:
+        1: punctuation.separator.decimal.ruby
+        2: storage.type.numeric.rational.ruby
```

Comment:
I thought about scoping these numbers as well with exectly the considderations you described in your comment.

I am kind of uncertain about it, but I tend to scope rational numbers as ordinary `float` for 3 reasons:

1. you did not use `rational` in the oct/hex/dec/bin or imaginary numbers above
2. rational numbers are a sub group of real/float numbers
3. rational is not yet part of the scope naming guideline

Reply:
I'm not sure if `float` is really the best choice for rational numbers, because the documentation explicitely highlights the difference between them:
> Rational numbers are exact, whereas Float numbers are inexact.
> 0.1r + 0.2r #=> (3/10)
> 0.1 + 0.2   #=> 0.30000000000000004

So rational numbers seem to be implemented as an own data type, just like complex numbers, and therefore I would prefer to give it a separate scope as well, to be consistent. To give both `12r` and `12.3r` the `constant.numeric.rational` scope, and not `constant.numeric.{integer|float}.rational` (dependant on the number part type) also corresponds with this.
> 12r         #=> (12/1)
> 12.3r       #=> (123/10)

For the numbers like `0xAAri` that fall in multiple categories, we have to choose the most relevant of them, because I think it wouldn't make sense to arbitrarily stack them into one large string (e.g. `constant.numeric.integer.hexadecimal.rational.imaginary`). This feels a bit similar to the recent discussion about `keyword.operator.logical` / `keyword.operator.word`. I think for numbers with base prefix the `hexadecimal`/`binary` part was the most important, for other notations including `i` the `complex.imaginary` property, and after that the `rational` property.

> 3. rational is not yet part of the scope naming guideline

I'm not sure which other languages have datatypes for rational numbers, but the only one from the default syntaxes I know of is Clojure, which uses `constant.numeric.ratio.clojure` for e.g. `2/3`, but I prefer `rational` more than `ratio` here. And I know that there are rational numbers `2//3` in Julia, but its syntax package doesn't even detect them. Perhaps it could be added to the guidelines?

Reply:
> To give both 12r and 12.3r the constant.numeric.rational scope

This is the compelling argument to use the `rational` scope here. None of the numbers is an integer, indeed.

> ...  I think it wouldn't make sense to arbitrarily stack them into one large string ...

Fully agree with you. This kind of long scope names doesn't make much sense as it is unable to be queried by selectors efficiently.

The issue with `complex.{real|imaginary}` and `int,float,rational,...` is just that these are two different kinds of perspective. We don't currently handle complex numbers completely in manner of scope naming guidelines.

To do so we would need to scope both the `real` and the `imaginary` part, while both of them can consist of `float` or `rational` numbers.

We would need to implement some thing like :

```
 12.0+10.4i
#^^^^ constant.numeric.complex.real constant.numeric.float
# ^ punctuation.separator.decimal
#    ^^^^^^ constant.numeric.complex.imaginary constant.numeric.float
#    ^ punctuation.separator.complex
#       ^ punctuation.separator.decimal
```

But this is nothing anyone considdered so far.

Reply:
And to add, in my opinion none of it should be `float` either, because from a mathematical point of view floating point numbers are a strict subset of rational numbers. The Ruby implementation seems to be a bit confusing though, because when tested for equality, the rational numbers apparently are mapped to floats (sorry if I went a bit too far with all these long numbers for this discussion :stuck_out_tongue_winking_eye:):
```ruby
r = 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001r
f = 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001
puts r
puts f
puts r == f
puts 10*r == 10*f
```
output (from https://repl.it, ruby 2.5.5p157 (2019-03-15 revision 67260) [x86_64-linux]):
```ruby
1/1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
0.0
true
false
```
But this actually applies more to the previously mentioned languages, which have rational numbers with arbitrary denominator like `1/3`, that itself can't be mapped to floating point numbers.

I agree with your considerations about complex numbers, but I don't think it would bring much benefit. It probably would make syntax definitions for numbers unnecessarily complicated and I don't think that any color scheme targets such specific scopes anyways, or that it would result in different highlighting. And it could be problematic to parse and to justify in what cases the `real` part and the `+` should be considered as its own number/operator, or to be part of the complex number.

Reply:
> ... but I don't think it would bring much benefit ... would make syntax definitions for numbers unnecessarily complicated

That's true. The question the otherway round than is: If we don't target `complex.real`, does it make sense to scope the imaginary part as `complex.imaginary` then? 

I just wonder about it, with a look into the other syntaxes, too.

Would it make sense to just scope imaginary numbers as normal `float|rational` with just the `i` scoped as `storate.type.imaginary` then?

Reply:
Well, I would say in cases where the language has a builtin datatype for complex numbers, `complex.imaginary` makes sense, and I like how the third-level part of the scope `{integer|float|complex|rational}` is always an indicator of the used datatype then. However, my proposal for rational/imaginary numbers with additional base prefix like `0xAAri` to be just `constant.numeric.integer.hexadecimal` isn't really accurate from that point of view, but I don't think there exists a perfect solution for this, unless we decide for repeatings like `constant.numeric.rational.ruby constant.numeric.complex.imaginary.ruby`. But I would rather avoid that and keep it simple, even if it isn't 100% accurate.

Reply:
I'm not heavily versed in number theory, but from doing a little reading, it seems that for complex numbers we are talking about numbers in the form, `a+bi` where `a` and `b` are real numbers and `i` is imaginary.

From this, I think the scoping should be:

```
 12.0+10.4i
#^^^^^^^^^^ constant.numeric.complex
#^^^^ constant.numeric.complex.real
#  ^ punctuation.separator.decimal
#     ^^^^ constant.numeric.complex.real
#    ^ punctuation.separator.complex
#       ^ punctuation.separator.decimal
#.        ^ constant.numeric.complex.imaginary
```

This way the entire thing is `constant.numeric.complex`. I'm not sure there is much value in trying to get super-specific beyond this level. The primary consumer of these scopes is color schemes, and the key thing we want to highlight here is that the `+` is not an addition sign, but part of making the whole thing an imaginary number.

Consistency is good, and I've got some work to do in terms of codifying various scope names we've been standardizing on into the scope naming guidelines, but I'd rather not let perfect get in the way of getting these changes in the hands of users.

Reply:
In `a+bi`, `a` is the real part and `b` is the imaginary part, while `i` (or sometimes `j`) is just a label for the imaginary unit. This is the important part here, so the scoping for `a` and `b` certainly has to be
```
 12.0+10.4i
#^^^^ constant.numeric.complex.real
#     ^^^^ constant.numeric.complex.imaginary
```
I'm not sure if there is much value in scoping the first part and the `+` sign at all, because a single `10.4i` will automatically parsed into `(0+10.4i)`, and the problem is to decide how to scope something like
```
12.0 +          10.4i
```
then. I would say this is the addition of a real and an imaginary number, and the `+` should be `keyword.operator`, but conventions might vary between languages.

Reply:
The `+` here is part of the imaginary part as it describes the direction of the vector. Complex numbers describe a vector in an coordinate system where the real part is the X and the imaginary part the Y (in an X,Y coordinate system).

Reply:
The most important part about this is operator precedence, because the plus operator (or separator, in a semantical sense) is subject to that and syntax definitions will struggle with something like `2*2+1i`.

Reply:
Okay, I was going off of https://en.wikipedia.org/wiki/Imaginary_number

> An imaginary number is a complex number that can be written as a real number multiplied by the imaginary unit *I*

However, after considering my separate comments about `3i` below, I'm thinking the whole thing `10.4i` or `12.0+10.4i` needs to be marked as imaginary. Based on the definition from Wikipedia (which may be imprecise for brevity), it seems an imaginary number is always a complex number. Based on this, I'm thinking as soon as `i` is present we need to mark it as imaginary, whether the form is "complex" or `3ri`, etc.

Reply:
In terms of whitespace between the first part and the separator, I would only scope it all as imaginary if there is no whitespace. But I'm not a user of imaginary numbers, so my preference should be pretty low in deciding what to do.